### PR TITLE
[Replicate][Test] tests that pp model grads are the same as single-device model grads

### PR DESCRIPTION
**Pull request by anshul-si**
**Summary:** Created a test so that we can verify that a model that has been pipelined + replicated has the same gradients as a reference model. To do this, I mapped the layers and their parameters in each partial model to the original full model and then compared the gradients. 
**Test Case**
1. pytest test/distributed/_composable/test_composability/test_pp_composability.py -k test_replicate_pp_grads

Stack from [ghstack](https://github.com/ezyang/ghstack) (oldest at bottom):
* __->__ #164890

cc @H-Huang @awgu @wanchaol @fegin @fduwjj @wz337 @wconstab @d4l3k @pragupta @msaroufim @dcci

### CODE REVIEW

File: test/distributed/_composable/test_composability/test_pp_composability.py [hunk 1]
```diff
@@ -604,6 +604,284 @@ def apply_same_precision(partial_model):
 
         torch.distributed.destroy_process_group()
 
+    @requires_accelerator_dist_backend(["nccl", "xccl"])
+    @skip_if_lt_x_gpu(8)
```
Comment (H-Huang) [inline test/distributed/_composable/test_composability/test_pp_composability.py]: this test works with just 4 GPUs

File: test/distributed/_composable/test_composability/test_pp_composability.py [hunk 2]
```diff
@@ -604,6 +604,284 @@ def apply_same_precision(partial_model):
...
+    @skip_but_pass_in_sandcastle_if(
+        not TEST_MULTIGPU and not TEST_XPU, "Test requires 8+ GPUs"
```
Comment (H-Huang) [inline test/distributed/_composable/test_composability/test_pp_composability.py:609]: same comment as above, edit message

File: test/distributed/_composable/test_composability/test_pp_composability.py [hunk 3]
```diff
@@ -604,6 +604,284 @@ def apply_same_precision(partial_model):
...
+        [
+            ScheduleGPipe,
+            Schedule1F1B,
+            ScheduleInterleaved1F1B,
+            ScheduleLoopedBFS,
+            ScheduleInterleavedZeroBubble,
+        ],
+    )
+    def test_replicate_pp_grads(self, ScheduleClass):
+        _device_raii = torch.device(device_type, self.device)
```
Comment (H-Huang) [inline test/distributed/_composable/test_composability/test_pp_composability.py]: confused how this is used? My guess based on the variable name is keep a reference to the device to keep it alive, but not sure why it is needed. If you can find out, i am curious!

File: test/distributed/_composable/test_composability/test_pp_composability.py [hunk 4]
```diff
@@ -604,6 +604,284 @@ def apply_same_precision(partial_model):
...
+                            rtol=1e-4,
+                            atol=1e-5,
+                        )
+                    else:
+                        assert pipeline_parameter.grad is None
+
+        pipeline_model_parameter_dict = {}
+
+        # Attach to a schedule
+        if issubclass(ScheduleClass, PipelineScheduleSingle):
```
Comment (H-Huang) [inline test/distributed/_composable/test_composability/test_pp_composability.py:740]: dang, this if-else needed for single stage schedules vs multistage schedules is so ugly, i should fix it 😭

File: test/distributed/_composable/test_composability/test_pp_composability.py [hunk 5]
```diff
@@ -604,6 +604,284 @@ def apply_same_precision(partial_model):
...
+                # For non-last stages, still run ref model to generate gradients
+                simulate_stage_forward_backward(ref_model, inputs, None)
+                simulate_all_reduce_grads(ref_model, dp_group)
+
+            # Step optimizers
+            for optimizer in optimizers:
+                optimizer.step()
+            ref_optimizer.step()
+
+            # torch.distributed.breakpoint()
```
Comment (H-Huang) [inline test/distributed/_composable/test_composability/test_pp_composability.py]: remove

**Review by H-Huang — APPROVED** [review]
LGTM, just some small comments / questions. Feel free to land it when you are ready!

File: test/distributed/_composable/test_composability/test_pp_composability.py (hunk 3 above)
Comment (anshul-si) [reply test/distributed/_composable/test_composability/test_pp_composability.py]: I believe its intended purpose was to do what you're saying. Turns out that its not needed in this instance so I will get rid of them.
